p2p/src/cybershake.rs

```diff
@@ -218,73 +236,246 @@ where
+        while written != msg.len() {
+            match self.write(&msg[written..]).await {
+                Ok(n) => written += n,
+                Err(e) => {
+                    return match e.kind() {
+                        io::ErrorKind::InvalidData => Err(Error::ProtocolError),
+                        _ => Err(Error::IoError(e)),
+                    }
+                }
+            };
+        }
+        self.flush().await.map_err(|e| Error::IoError(e))?;
+        Ok(())
+    }
+}
+
+impl<W: AsyncWrite + Unpin> Outgoing<W> {
+    fn cipher_buf(&mut self) {
+        self.flushing = true;
         self.kdf.append_u64(b"seq", self.seq);
         let mut key = [0u8; 32];
         self.kdf.challenge_bytes(b"key", &mut key);
 
         let ad = encode_u64le(self.seq);
 
         let ciphertext = Aes128PmacSiv::new(GenericArray::clone_from_slice(&key))
-            .encrypt(&[&ad], msg)
-            .map_err(|_| Error::ProtocolError)?;
+            .encrypt(&[&ad], &self.buf)
+            .map_err(|_| unimplemented!())
+            .unwrap();
+
+        self.buf.clear();
+
+        // Write length of the message.
+        Write::write(&mut self.buf, &encode_u16le(ciphertext.len() as u16)[..]).unwrap();
+        // Write a message.
+        Write::write(&mut self.buf, &ciphertext).unwrap();
 
         self.seq += 1;
+    }
 
-        // Write the length prefix and the ciphertext.
-        self.writer
-            .write(&encode_u64le(ciphertext.len() as u64)[..])
-            .await?;
-        self.writer.write(&ciphertext[..]).await?;
-        self.writer.flush().await?;
-        Ok(())
+    fn flush_pending_ciphertext(&mut self, cx: &mut Context) -> Poll<Result<(), io::Error>> {
+        loop {
+            let poll = self
+                .writer
+                .as_mut()
+                .poll_write(cx, &self.buf[self.ciphertext_sent..]);
+            let n = ready!(poll);
+            self.ciphertext_sent += n;
+            if self.ciphertext_sent == self.buf.len() {
+                self.ciphertext_sent = 0;
+                self.flushing = false;
+                self.buf.clear();
+                return Poll::Ready(Ok(()));
+            }
+        }
     }
 }
 
-impl<W: AsyncRead + Unpin> Incoming<W> {
-    pub async fn receive_message(&mut self) -> Result<Vec<u8>, Error> {
-        let mut lenbuf = [0u8; 8];
-        let seq = self.seq;
-        self.seq += 1;
-        self.reader.read_exact(&mut lenbuf[..]).await?;
-        let len = LittleEndian::read_u64(&lenbuf) as usize;
+impl<W: AsyncWrite + Unpin> AsyncWrite for Outgoing<W> {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        let me = self.get_mut();
 
-        // length must include IV prefix (16 bytes)
-        if len < 16 {
-            return Err(Error::ProtocolError);
+        if me.flushing {
+            ready!(me.flush_pending_ciphertext(cx));
```

Comment:
...then, here i'd make flushing unconditional: if we have some CT to send, we need to flush it. The necessary checks will be inside the method.

Reply:
Checks were moved into `flush_pending_ciphertext()`